connectors/yarpc/yarpc.go

```diff
@@ -478,29 +478,41 @@ func (c *Connector) Scan(ctx context.Context, ei *dosa.EntityInfo, minimumFields
 // CheckSchema is one way to register a set of entities. This can be further validated by
 // a schema service downstream.
 func (c *Connector) CheckSchema(ctx context.Context, scope, namePrefix string, eds []*dosa.EntityDefinition) (int32, error) {
-	// convert the client EntityDefinition to the RPC EntityDefinition
-	rpcEntityDefinition := make([]*dosarpc.EntityDefinition, len(eds))
-	for i, ed := range eds {
-		rpcEntityDefinition[i] = EntityDefinitionToThrift(ed)
+	response, err := c.checkSchema(ctx, scope, namePrefix, eds, false)
+	if err != nil {
+		return dosa.InvalidVersion, wrapError(err, "CheckSchema failed", scope, c.Config.ServiceName)
 	}
-	csr := dosarpc.CheckSchemaRequest{EntityDefs: rpcEntityDefinition, Scope: &scope, NamePrefix: &namePrefix}
 
-	response, err := c.Client.CheckSchema(ctx, &csr, VersionHeader())
+	return *response.Version, nil
+}
 
+// CheckSchemaToUpsert checks whether the provided entities are compatible with the latest applied schema.
+// A non-nil error indicates the entities are not backward-compatible with the latest schema, thus will fail
+// if they were to be upserted.
+func (c *Connector) CheckSchemaToUpsert(ctx context.Context, scope, namePrefix string, eds []*dosa.EntityDefinition) (int32, error) {
+	response, err := c.checkSchema(ctx, scope, namePrefix, eds, true)
 	if err != nil {
-		return dosa.InvalidVersion, wrapError(err, "CheckSchema failed", scope, c.Config.ServiceName)
+		return dosa.InvalidVersion, wrapError(err, "CheckSchema compatibility failed", scope, c.Config.ServiceName)
 	}
 
 	return *response.Version, nil
 }
 
+func (c *Connector) checkSchema(ctx context.Context, scope, namePrefix string, eds []*dosa.EntityDefinition, toUpsert bool) (*dosarpc.CheckSchemaResponse, error) {
```

Comment:
See my comments in https://github.com/uber/dosa-idl/pull/22. If we make `CheckSchema` and `CheckSchemaToUpsert` two different calls, then this function probably shouldn't exist.

Reply:
right. Also I've changed the https://github.com/uber/dosa-idl/pull/22 to use two rpc calls